tasks.py

```diff
@@ -0,0 +1,154 @@
+"""
+tasks.py defines tasks to run via ssh on one of the defined hosts.
+The functions with the @task decoration can be executed from the cli
+by calling `invoke task arg1 ... argn`, e.g. `invoke backup staging`.
+The caller has to be set up to make ssh connections to the intended host
+as the defined user.
+
+## Defined tasks
+* backup env
+    starting a routine to create an encrypted backup of the database
+    in the environment `env`.
+* deploy env github_user github_token
+    downloading the most recent image and code for the given environment,
+    then migrating the database if any migrations are necessary,
+    finally restarting the docker containers. The github user and token are
+    used to download the docker image from the github containe registry.
+
+
+"""
+import datetime
+import io
+import os
+from dataclasses import dataclass
+
+from fabric import Connection
+from invoke import task
+
+
+@dataclass
+class HostConfig:
+    host_name: str
+    git_branch: str
+    docker_version: str
+    github_user: str
+    github_token: str
+    backup_name: str
+    user: str = "deploy-web"
+    code_dir: str = "~/OpenOversight"
+    backup_dir: str = "~/openoversight_backup/"
+    image_path: str = "ghcr.io/lucyparsons/openoversight"
+    key_id: str = "OpenOversight"
+
+    def get_connection(self):
+        return Connection(host=self.host_name, user=self.user)
+
+
+def get_configs(env, github_user=None, github_token=None) -> HostConfig:
+    configs = {
+        "staging": HostConfig(
+            host_name="staging.openoversight.com",
+            git_branch="develop",
+            docker_version="latest",
+            github_user=github_user,
+            github_token=github_token,
+            backup_name="backup_staging",
+        ),
+        "prod": HostConfig(
+            host_name="openoversight.com",
+            git_branch="main",
+            docker_version="stable",
+            github_user=github_user,
+            github_token=github_token,
+            backup_name="backup_prod",
+        ),
+    }
+    return configs[env]
+
+
+def change_code(c, config):
+    c.run(f"git fetch origin {config.git_branch}")
+    c.run("git reset --hard FETCH_HEAD")
+    c.run("git clean -df")
+
+
+def migrate(c, config):
+    c.run(
+        f"DB_IMAGE_TAG={config.docker_version} docker compose -f docker-compose.prod-img.yml run --rm --no-deps web flask db upgrade",
+        echo=True,
+    )
+
+
+def deploy_(c: Connection, config: HostConfig):
+    print("Start new deployment.")
```

Comment:
I know it's a little bit more work, but I'd recommend converting these from `print` to `log.info` statements and setting up a `basicConfig` to log at INFO level by default. This would help provide timestamps for when each statement is issued (I've found myself shaking my fist at myself in the past for not having timestamps when using `print` 😅).

Reply:
Thank you, I agree that it is very important to have. Unfortunately python makes logging so weird by default (time is parsed with a comma instead of decimal for microseconds and in local time), so there are some extra lines to make it acceptable.